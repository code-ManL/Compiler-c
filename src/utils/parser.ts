
var keywords: {
  [k: string]: any
} = {
  "char": 101,
  "int": 102,
  "float": 103,
  "break": 104,
  "const": 105,
  "return": 106,
  "void": 107,
  "continue": 108,
  "do": 109,
  "while": 110,
  "if": 111,
  "else": 112,
  "for": 113,
  "function": 114
}



var separators = {
  "{": 301,
  "}": 302,
  ";": 303,
  ",": 304,
  "(": 201,
  ")": 202,
  "[": 203,
  "]": 204,
}


var operators = {
  "!": 205,
  "*": 206,
  "/": 207,
  "%": 208,
  "+": 209,
  "-": 210,
  "<": 211,
  "<=": 212,
  ">": 213,
  ">=": 214,
  "==": 215,
  "!=": 216,
  "&&": 217,
  "||": 218,
  "=": 219,
  ".": 220
}

const state = {
  "error": -1,
  "initial": 0,
  "charOpen": 1,
  "charEnd": 2,
  "numOpen": 3,
  "numEnd": 4,
  "separators": 5,
  "operators": 6,
  "logOpen": 7,
  "logEnd": 8,
  "fnOpen": 9,
  "putOpen": 10,
  "putEnd": 11,
  "fnCharOpen": 12,
  "fnCharEnd": 13,
}

function isAlpha(c: string) {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || '_'
}

function isDigit(d: string) {
  return d >= '0' && d <= '9'
}

function isSpace(s: string) {
  return /s/.test(s)
}


type Ttoken = {
  type: string,
  state: number,
  value: string
}



const punctuator = ["{", "}", ";", ",", "(", ")", "[", "]"]
const keyword = ['function', 'var', 'let', 'const']



export function tokenize(s: string) {
  let i = 0;
  let token: Ttoken[] = []
  const chars: string[] = []
  let currentState = state.initial

  while (s) {
    const c = s[i]
    switch (currentState) {
      case state.initial:
        if (isAlpha(c)) {
          currentState = state.charOpen
          chars.push(c)
          s = s.slice(1)
        }
        else if (isDigit(c)) {
          currentState = state.numOpen
          chars.push(c)
          s = s.slice(1)
        }
        else if (isSpace(c)) {
          s = s.slice(1)
        }
        break
      case state.charOpen:
        if (isAlpha(c)) {
          chars.push(c)
          s = s.slice(1)
        }
        else if (isDigit(c)) {
          const c = chars.join('')
          token.push({
            type: 'Identifier',
            state: 2,
            value: c
          })
          chars.length = 0
          currentState = state.numOpen
        }
        else if (isSpace(c)) {
          const c = chars.join('')
          if (keyword.includes(c)) {
            token.push({
              type: 'Keyword',
              state: 1,
              value: c
            })
          }
          else {
            token.push({
              type: 'Identifier',
              state: 2,
              value: c
            })
          }
          chars.length = 0
          s = s.slice(1)
          currentState = state.initial
        }
        else if (punctuator.includes(c)) {
          const c = chars.join('')
          if (keyword.includes(c)) {
            token.push({
              type: 'Keyword',
              state: 1,
              value: c
            })
          }
          else {
            token.push({
              type: 'Identifier',
              state: 2,
              value: c
            })
          }
          chars.length = 0
          currentState = state.putOpen
        }
        break
      case state.numOpen:
        if (isDigit(c)) {
          chars.push(c)
          s = s.slice(1)
        } else if (isAlpha(c) || isSpace(c)) {
          const c = chars.join('')
          token.push({
            type: 'Number',
            state: 3,
            value: c
          })
          chars.length = 0
          currentState = state.charOpen
        }
        break
      case state.putOpen:
        if (punctuator.includes(c)) {
          token.push({
            type: 'Punctuator',
            state: 4,
            value: c
          })
          s = s.slice(1)
        }
        else if (isAlpha(c)) {
          currentState = state.charOpen
        }
        else if (isDigit(c)) {
          currentState = state.numOpen
        }
    }
    i++
  }
}

