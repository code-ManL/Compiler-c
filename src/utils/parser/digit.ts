enum Digit {
  PLUS = 209

}


export { }


