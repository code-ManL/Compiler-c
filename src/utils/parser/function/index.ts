
/**
 * 函数声明
 * @default: <函数声明> => <标识符>(<函数声明形参列表>)
 * 
 */
function functionDeclareSentence() {

}
export var functionDeclareSentence_first:string[] = []
export var functionDeclareSentence_follow:string[] = []

/**
 * 函数声明形参列表
 * @default: <函数声明形参列表> => <函数声明形参> | 3
 */
function functionArgumnetsTableDeclareSentence() {

}

/**
 * 函数声明形参
 * @default: <函数声明形参> => <变量类型> | <变量类型>,<函数声明形参>
 * @description: 消除回溯
 *    <函数声明形参> => <变量类型><函数声明形参 Tail>
 *    <函数声明形参 Tail> => ,<函数声明形参> | 3
 */
function functionArgumnetsDeclareSentence() { 

}


export {
  functionArgumnetsDeclareSentence,
  functionArgumnetsTableDeclareSentence,
  functionDeclareSentence
}
