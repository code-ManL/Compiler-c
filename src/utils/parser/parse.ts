

function clearnWs(code: string) {
  for (let i = 0; i < code.length; i++) {
    console.log(code[i] === '\n');

    console.log(code[i]);
  }
}


function isAlpha(c: string) {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}

function isDigit(v: string) {
  return v >= '0' && v <= '9'
}


export function parser(code: string) {
  
  

}