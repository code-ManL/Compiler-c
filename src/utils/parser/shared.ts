import type { Ttoken } from '../tokensize/types'

// mock tokens
var tokens: Ttoken[] = [
  {
    type: '',
    state: 1,
    value: '',
    col: 1,
    row: 2,
    start: 1,
  }
]


// 报错
export function toThrowError(error: string) {
  console.log(error);
}

// 获取下一个token
export function getNextToken() {
  return tokens.shift() || { value: 'end' }
}
