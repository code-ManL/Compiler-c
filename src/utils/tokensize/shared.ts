import keyword from "./keyowrd"
import operators from './operators'
import punctuator from './punctuator'


function isAlpha(c: string) {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c === '_'
}

function isDigit(d: string) {
  return d >= '0' && d <= '9'
}

function isSpace(s: string) {
  return s === ' '
}

function isPunctuator(c: string) {
  return punctuator.includes(c)
}

function isOperators(c: string) {
  return operators.includes(c)
}

function isKeyword(c: string) {
  return keyword.includes(c)
}

function isOct(c: string) {
  return c === 'x' || c === 'X'
}

function isHex(c: string) {
  return (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
}

export {
  isAlpha,
  isDigit,
  isSpace,
  isPunctuator,
  isOperators,
  isKeyword,
  isOct,
  isHex
}
