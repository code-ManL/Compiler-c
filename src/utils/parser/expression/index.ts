export * from './assignExpresstion'
export * from './booleanExpreestion'
export * from './computeExpressiton'
export * from './relationExpresstion'

import { getNextToken, toThrowError } from '../shared'
import { assgineExpression } from './assignExpresstion'
import { booleanExpression } from './booleanExpreestion'
import { computedExpression } from './computeExpressiton'
import { relationExpression } from './relationExpresstion'
/**
 * <表达式> => <算数表达式> | <关系表达式> | <布尔表达式> | <赋值表达式>
 */

export function expresstion() {
}







