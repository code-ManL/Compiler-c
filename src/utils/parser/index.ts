import type { Ttoken } from '../tokensize/types'
import { getPre, getNextToken, toThrowError, getFirst, getCurrentToken, addPre } from './shared'
import { programEntry } from '../parser/program/index'
import { declareSentenceTable } from './sentence'



// <程序> -> <声明语句列表> <MAIN函数定义> <函数列表>
export function parser() {
  let token = getCurrentToken()

  

  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  




  
  
  
  
  
  /** 
  if (getFirst("<程序>").includes(token.value)) {
    if (getFirst("<声明语句列表>")) {
      declareSentenceTable()
    } else if (getFirst("<MAIN函数定义>")) {


    } else {


    }
  }
  else {
    toThrowError('不符合程序文法')
    return
  }

  */
}


