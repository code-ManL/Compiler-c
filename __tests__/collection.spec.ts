import { getFirst, getFollow, target } from '../src/utils/parser/shared'

describe('first', async () => {
  test('frist', () => {
    for (const key of Object.keys(target)) {
      expect(getFirst(key)).toMatchSnapshot()
    }

  })
})


describe('follow', async () => {
  test('follow', () => {
    for (const key of Object.keys(target)) {
      expect(getFollow(key)).toMatchSnapshot()
    }
  })
})
