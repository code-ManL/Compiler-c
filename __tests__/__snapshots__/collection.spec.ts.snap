// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`first > frist 1`] = `
[
  "(",
  "i",
]
`;

exports[`first > frist 2`] = `
[
  "+",
  "None",
]
`;

exports[`first > frist 3`] = `
[
  "(",
  "i",
]
`;

exports[`first > frist 4`] = `
[
  "*",
  "None",
]
`;

exports[`first > frist 5`] = `
[
  "(",
  "i",
]
`;

exports[`follow > follow 1`] = `
[
  "#",
  ")",
]
`;

exports[`follow > follow 2`] = `
[
  "#",
  ")",
]
`;

exports[`follow > follow 3`] = `
[
  "+",
  "#",
  ")",
]
`;

exports[`follow > follow 4`] = `
[
  "+",
  "#",
  ")",
]
`;

exports[`follow > follow 5`] = `
[
  "*",
  "+",
  "#",
  ")",
]
`;
